File: String/28. Implement strStr().cpp
/*
28. Implement strStr()
Two Pointers, String

Description:
Implement strStr().
Return the index of the first occurrence of needle in haystack, or -1 if needle is not part of haystack.
Clarification:
What should we return when needle is an empty string? This is a great question to ask during an interview.
For the purpose of this problem, we will return 0 when needle is an empty string. This is consistent to C's strstr() and Java's indexOf().


Example 1:
Input: haystack = "hello", needle = "ll"
Output: 2

Example 2:
Input: haystack = "aaaaa", needle = "bba"
Output: -1

Example 3:
Input: haystack = "", needle = ""
Output: 0

Similar Questions:
Shortest Palindrome - Hard
Repeated Substring Pattern - Easy
*/

// Method 1: Sliding Window
class Solution {
public:
    int strStr(string haystack, string needle) {
        if(needle.size()==0){
            return 0;
        }else if(haystack.size()==0){
            return -1;
        }
        int right = 0, left = 0;
        unordered_map<char, int> m_needle;
        for(auto& c:needle){
            m_needle[c]++;
        }
        unordered_map<char, int> m_window; int match = 0;
        while(right < haystack.size()){
            char c = haystack[right];
            if(m_needle.find(c)!=m_needle.end()){
                m_window[c]++;
                if(m_window[c]==m_needle[c]){
                    match++;
                }
            }
            right++;
            while(match==m_needle.size()){
                char cl = haystack[left];
                if(m_window.find(cl)!=m_window.end()){
                    m_window[cl]--;
                    if(m_window[cl]<m_needle[cl]){
                        match--;
                    }
                }
                string str_window = haystack.substr(left, right-left);
                if(str_window == needle){
                    return left;
                }
                left++;
            }
            
        }
        return -1;
    }
};


// Method 2: string.find(substr) searches the string for the first occurrence of the sequence specified by its        arguments. 
class Solution {
public:
    int strStr(string haystack, string needle) {
        if(needle.empty()) return 0;
        return haystack.find(needle);
    }
};

File: Array/28. Implement strStr().py
# Method 1: string.find(substr)
class Solution(object):
    def strStr(self, haystack, needle):
        """
        :type haystack: str
        :type needle: str
        :rtype: int
        """
        if len(needle) == 0:
            return 0;
        return haystack.find(needle)


'''
Method 2: Sliding Window
          Note: 1. create the dictionaries as defaultdic(int)
                2. check if key exists without updating the value, use if key in Dic
'''
class Solution(object):
    def strStr(self, haystack, needle):
        """
        :type haystack: str
        :type needle: str
        :rtype: int
        """
        if len(needle) == 0:
            return 0
        Dic_needle, Dic_window = defaultdict(int), defaultdict(int)
        match, r, l = 0, 0, 0
        for c in needle:
            Dic_needle[c] += 1
        
        while r < len(haystack):
            c = haystack[r]
            if c in Dic_needle:
                Dic_window[c] += 1
                if Dic_window[c] == Dic_needle[c]:
                    match += 1
            r += 1
            while match == len(Dic_needle):
                if haystack[l:r] == needle:
                    return l
                cl = haystack[l]
                if cl in Dic_needle:
                    Dic_window[cl] -= 1
                    if Dic_window[cl] < Dic_needle[cl]:
                        match -= 1
                l += 1
        
        return -1
